myst_nbs/case_studies/reinforcement_learning.myst.md

```diff
@@ -0,0 +1,541 @@
+    Q = np.array([0.5, 0.5])
+    for i in range(n):
+        # Apply the Softmax transformation
+        exp_Q = np.exp(beta * Q)
+        prob_a = exp_Q / np.sum(exp_Q)
+
+        # Simulate choice and reward
+        a = rng.choice([0, 1], p=prob_a)
+        r = rng.random() < p_r[a]
+
+        # Update Q table
+        Q[a] = Q[a] + alpha * (r - Q[a])
+
+        # Store values
+        actions[i] = a
+        rewards[i] = r
+        Qs[i] = Q.copy()
+
+    return actions, rewards, Qs
+```
+
+```{code-cell} ipython3
+:id: ceNagbmsZXW6
+
+true_alpha = 0.5
+true_beta = 5
+n = 150
+actions, rewards, Qs = generate_data(rng, true_alpha, true_beta, n)
+```
+
+```{code-cell} ipython3
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 208
+id: MDhJI8vOXZeU
+outputId: 60f7ee37-2d1f-44ad-afff-b9ba7d82a8d8
+---
+_, ax = plt.subplots(figsize=(12, 5))
+x = np.arange(len(actions))
+
+ax.plot(x, Qs[:, 0] - 0.5 + 0, c="C0", lw=3, alpha=0.3)
+ax.plot(x, Qs[:, 1] - 0.5 + 1, c="C1", lw=3, alpha=0.3)
+
+s = 7
+lw = 2
+
+cond = (actions == 0) & (rewards == 0)
+ax.plot(x[cond], actions[cond], "o", ms=s, mfc="None", mec="C0", mew=lw)
+
+cond = (actions == 0) & (rewards == 1)
+ax.plot(x[cond], actions[cond], "o", ms=s, mfc="C0", mec="C0", mew=lw)
+
+cond = (actions == 1) & (rewards == 0)
+ax.plot(x[cond], actions[cond], "o", ms=s, mfc="None", mec="C1", mew=lw)
+
+cond = (actions == 1) & (rewards == 1)
+ax.plot(x[cond], actions[cond], "o", ms=s, mfc="C1", mec="C1", mew=lw)
+
+ax.set_yticks([0, 1], ["left", "right"])
+ax.set_ylim(-1, 2)
+ax.set_ylabel("action")
+ax.set_xlabel("trial")
+
+ax.scatter(0, 20, c="k", s=50, lw=lw, label="Reward")
+ax.scatter(0, 20, c="w", ec="k", s=50, lw=lw, label="No reward")
+ax.plot([0, 1], [20, 20], c="k", lw=3, alpha=0.3, label="Qvalue (centered)")
+
+handles, labels = ax.get_legend_handles_labels()
+order = (1, 2, 0)
+handles = [handles[idx] for idx in order]
+labels = [labels[idx] for idx in order]
+ax.legend(handles, labels, fontsize=12, loc=(1.01, 0.27));
+```
+
++++ {"id": "6RNLAtqDXgG_"}
+
+The plot above shows a simulated run of 150 trials, with parameters $\alpha = .5$ and $\beta = 5$, and constant reward probabilities of $.4$ and $.6$ for the left (blue) and right (orange) actions, respectively. 
+
+Solid and empty dots indicate actions followed by rewards and no-rewards, respectively. The solid line shows the estimated $Q$ value for each action centered around the respective colored dots (the line is above its dots when the respective $Q$ value is above $.5$, and below otherwise). It can be seen that this value increases with rewards (solid dots) and decreases with non-rewards (empty dots). 
+
+The change in line height following each outcome is directly related to the $\alpha$ pamater. The influence of the $\beta$ parameter is more difficult to grasp, but one way to think about it is that the higher its value, the higher the probability of flipping between actions regardless of their estimated values. 
```

Comment:
There's a mistake in this paragraph. The higher the `beta` the more deterministic the responses are / or the more sensitive they are to the observed rewards. I think flipping the adjective does not quite convey this information...

```suggestion
The change in line height following each outcome is directly related to the $\alpha$ parameter. The influence of the $\beta$ parameter is more difficult to grasp, but one way to think about it is that the higher its value, the more an agent will stick to the action that has the highest estimated value, even if the difference between the two is quite small. Conversely, as this value approaches zero, the agent will start picking randomly between the two actions, regardless of their estimated values.
```

Reply:
Thanks! I added it in https://github.com/pymc-devs/pymc-examples/pull/410/commits/5252c2b4d8304d20b7eb5daf2ccafdc60e229f0c (I am always hesitant to commit suggestions in jupyter notebooks 😄 )